File: exercise/25/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (k == 1) {
            return head;
        }
        ListNode fake = ListNode(0);
        ListNode *point = &fake;
        point->next = head;
        while (true) {
            ListNode *last = point, *first;
            bool done = false;
            for (int i = 0; i < k; i++) {
                if (last->next) {
                    last = last->next;
                } else {
                    done = true;
                    break;
                }
            }
            if (done)
                break;
            for (int i = 0; i < k - 1; i++) {
                last = point->next;
                first = last->next;
                last->next = first->next;
                first->next = point->next;
                point->next = first;
            }
            point = last;
        }
        return fake.next;
    }
};

File: exercise/25/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if k == 1:
            return head
        fake = ListNode(0)
        fake.next = head
        head = fake
        while True:
            tail = head
            for i in range(k):
                tail = tail.next
                if not tail:
                    return fake.next
            for i in range(k - 1):
                head_next = head.next
                head.next = head_next.next
                head_next.next = tail.next
                tail.next = head_next
            for i in range(k):
                head = head.next
